Cargo.toml

```diff
@@ -25,7 +29,12 @@ mio_07 = {package="mio", version="0.7.0", features=["os-util", "uds"], optional=
 # examples and tests for mio_07 also requires mio feature os-poll,
 # but adding it as a dev-dependency would also enable it in all cases (cargo bug #4866)
 # instead RUSTFLAGS='--cfg feature="os-poll"' must be used to build & run mio_07 tests
+tokio_02 = {package="tokio", version = "0.2", features = ["io-driver"], optional=true}
+
+[target."cfg(unix)".dev-dependencies]
+tempdir = "0.3"
+tokio_02 = {package="tokio", version = "0.2", features = ["macros", "rt-core", "test-util"]}
```

Comment:
Ideally this would be `optional` as well, but I don't think Cargo allows optional dev-dependencies. Thus, it's either forcing these tokio test-related features for non-test builds, or forcing all test builds to take a tokio dependency.

Reply:
What do you think about moving the dev dependency features to RUSTFLAGS? If it's not pointless I can do that later.

Reply:
Interesting; I wasn't aware of this issue, or workaround. I tried moving the dev features to `RUSTFLAGS`, but had some issues with `macros` and tokio's failure to pull in the `tokio-macros` sub-crate:

```
$ RUSTFLAGS='--cfg feature="os-poll" --cfg feature="rt-core" --cfg feature="macros"' cargo test --all-features

...

error[E0432]: unresolved import `tokio_macros`
   --> /home/jon/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.22/src/lib.rs:403:13
    |
403 |     pub use tokio_macros::select_priv_declare_output_enum;
    |             ^^^^^^^^^^^^ use of undeclared type or module `tokio_macros`
```

(note, I realized that `test-util` isn't actually used, so that will be dropped regardless of the outcome of this)

One workaround is just importing `tokio-macros` itself into dev-dependencies, and then setting up tests such that:

```rust
#[tokio_macros::test]
async fn test_listener_accept() {
    ...
}
```

which isn't something I've really seen other crates do (especially in documentation, which exclusively uses `tokio::main`), but I couldn't come up with a better solution offhand.

Thoughts?

Reply:
There's also [`cargo-hack`](https://github.com/taiki-e/cargo-hack) which looks to address these limitations in cargo. I did a quick trial run and it didn't seem to like running test with `--all-features`, but worked as expected when I wrote everything out manually:

```
$ cargo hack --features="mio_07,tokio,tokio_02/macros,tokio_02/rt-core,mio_07/os-poll" test
```

Reply:
Thanks for the attempt. We'll just have to accept it being enabled for now.
(by the time I'm releasing v1.0 of this it will probably have been fixed in cargo :)

cargo-hack looks interesting but feels too complex.